apps/prairielearn/src/pages/instructorFileEditor/instructorFileEditor.js

```diff
@@ -8,301 +8,289 @@ import * as fs from 'fs-extra';
+        res.locals.course_owners = owners;
+        res.status(403).render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+      })
+      .catch((err) => next(err));
+    return;
   }
 
-  let workingPath;
-  if (req.params[0]) {
-    try {
-      workingPath = decodePath(req.params[0]);
-    } catch (err) {
-      return next(new Error(`Invalid path: ${req.params[0]}`));
-    }
-  } else {
-    return next(new Error(`No path`));
-  }
-
-  let fileEdit = {
-    uuid: uuidv4(),
-    userID: res.locals.user.user_id,
-    authnUserId: res.locals.authn_user.user_id,
-    courseID: res.locals.course.id,
-    coursePath: res.locals.course.path,
-    dirName: path.dirname(workingPath),
-    fileName: path.basename(workingPath),
-    fileNameForDisplay: path.normalize(workingPath),
-    aceMode: modelist.getModeForPath(workingPath).mode,
-  };
-
   // Do not allow users to edit the exampleCourse
   if (res.locals.course.example_course) {
-    return next(error.make(400, `attempting to edit file inside example course: ${workingPath}`));
+    res.status(403).render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+    return;
   }
 
-  // Do not allow users to edit files outside the course
-  const fullPath = path.join(fileEdit.coursePath, fileEdit.dirName, fileEdit.fileName);
-  const relPath = path.relative(fileEdit.coursePath, fullPath);
-  debug(
-    `Edit file in browser\n fileName: ${fileEdit.fileName}\n coursePath: ${fileEdit.coursePath}\n fullPath: ${fullPath}\n relPath: ${relPath}`,
-  );
-  if (!contains(fileEdit.coursePath, fullPath)) {
-    return next(
-      error.make(400, `attempting to edit file outside course directory: ${workingPath}`),
+  // Do not allow users to edit files in bad locations (e.g., outside the
+  // current course, outside the current course instance, etc.). Do this by
+  // wrapping everything in getPaths, which throws an error on a bad path.
+  getPaths(req, res, (err, paths) => {
+    if (ERR(err, next)) return;
+
+    // We could also check if the file exists, if the file actually is a
+    // file and not a directory, if the file is non-binary, etc., and try
+    // to give a graceful error message on the edit page rather than send
+    // the user to an error page.
+    //
+    // We won't do that, on the assumption that most users get to an edit
+    // page through our UI, which already tries to prevent letting users
+    // go where they should not.
+
+    const fullPath = paths.workingPath;
+    const relPath = paths.workingPathRelativeToCourse;
+    let fileEdit = {
+      uuid: uuidv4(),
+      userID: res.locals.user.user_id,
+      authnUserId: res.locals.authn_user.user_id,
+      courseID: res.locals.course.id,
+      coursePath: paths.coursePath,
+      dirName: path.dirname(relPath),
+      fileName: path.basename(relPath),
+      fileNameForDisplay: path.normalize(relPath),
+      aceMode: modelist.getModeForPath(relPath).mode,
+      jobSequence: /** @type {any} */ (null),
```

Comment:
This is a bandaid over a TypeScript issue that should really be resolved with a full rewrite of all this stuff, ideally to avoid this "god" object that we pass around to everything.